### PR TITLE
ENH: expand OpenMP utilities and move from denspeed.pyx to dipy.utils

### PR DESCRIPTION
**Pull request by grlee77**
I was recently working on adding OpenMP support to some functions in dipy.align and needed some utilities similar to the ones defined within `denoise/denspeed.pyx`.

I propose moving these to a common location `dipy.utils.omp` where they can be more easily found and reused.

The only change in functionality to `nlmeans` would be the default number of threads set when `num_threads` is not specified,but the user has also defined the `OMP_NUM_THREADS` environement variable.  Previously the default number of threads was always `openmp.omp_get_num_procs()`, but now it will try to default to the `OMP_NUM_THREADS` instead.  At least on my system, `omp_get_num_procs` returns the total number of possible threads including hyperthreading which actually makes things slower than just using the number of physical cores.

### CODE REVIEW

**Comment by riddhishb** [comment]
@grlee77 this is a good idea, just a quick question. 
Does this mean that when `nlmeans` is run without specifying num_threads, it would use all the cores, I believe thats what `OMP_NUM_THREADS` does right?
I am not clear on that part

**Comment by grlee77** [comment]
By default OMP_NUM_THREADS is probably not set on most systems unless the user has explicitly done so.  In that case the max number of threads returned by omp_get_num_procs would be used.

OMP_NUM_THREADS is an environment variable that can be set by the user (e.g. via `export OMP_NUM_THREADS=4` in a bash terminal on linux or osx prior to starting python).  For the laptop I am on now there are 4 physical CPU cores, but 8 maximum threads.  In practice I find using 4 threads is faster than 8 so I set OMP_NUM_THREADS=4 and would like dipy to honor that choice rather than always forcing to 8.

However, looking more closely at these docs (https://gcc.gnu.org/onlinedocs/libgomp/OMP_005fNUM_005fTHREADS.html), I see that technically OMP_NUM_THREADS can also be set to a comma separated list.  I have not supported that case here.  The code as implemented here would fail to convert the list to an integer and fallback to the maximum number of threads as returned by omp_get_num_procs.

**Comment by riddhishb** [comment]
Got it! :) and this is only when the num_threads keyword is not specified, right?

**Comment by grlee77** [comment]
Right.

**Comment by arokem** [comment]
@omarocegueda : would you mind giving this one a look as well? Is it still relevant?

File: dipy/denoise/denspeed.pyx [hunk 1]
```diff
@@ -117,8 +111,8 @@ def _nlmeans_3d(double[:, :, ::1] arr, double[:, :, ::1] mask,
 
                     out[i, j, k] = process_block(arr, i, j, k, B, P, sigma)
 
-    if have_openmp and num_threads is not None:
-        openmp.omp_set_num_threads(all_cores)
+    if num_threads is not None:
+        restore_default_num_threads()
```
Comment (omarocegueda) [inline dipy/denoise/denspeed.pyx]: Just out of curiosity, is it possible that this logic has unexpected side effects? What I understand is that we are "restoring" the number of threads after finishing the parallel block. But what if the number of threads was not the default when we called this function?

Comment (omarocegueda) [inline dipy/denoise/denspeed.pyx]: Yes, I think that would have been better. Resetting the number of threads to the default seems to me as unexpected as not restoring it at all. Anyway, that issue is unrelated to this PR =).

**Comment by arokem** [comment]
Other than your previous comment, @omarocegueda : what do you think here?

**Comment by arokem** [comment]
OK - this seems ready to go, so unless someone pipes up, I will merge it in a couple of days.
